### PR TITLE
Fix and Add Logging for StateMachine websocket close

### PR DESCRIPTION
**Pull request by shanedewael**
###  Summary

Addresses #531 and #527. Adds flags to `KeepAlive` and logs in client on unexpected websocket close during `handshaking` state. Also addresses race condition that could cause this unexpected close.

### Requirements (place an `x` in each `[ ]`)

* [x] I've read and understood the [Contributing Guidelines](https://github.com/slackapi/node-slack-sdk/blob/master/.github/contributing.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).

### CODE REVIEW

File: src/RTMClient.ts [hunk 1]
```diff
@@ -145,6 +145,22 @@ export class RTMClient extends EventEmitter {
...
+                this.logger.debug(`reconnecting after unexpected close ${context.eventPayload.reason} 
+                  ${context.eventPayload.code} with isMonitoring set to ${this.keepAlive.isMonitoring} 
+                  and recommendReconnect set to ${this.keepAlive.recommendReconnect}`);
+              })
+              .transitionTo('disconnected')
+              .withAction((_from, _to, context) => {
+                this.logger.debug(`disconnected after unexpected close ${context.eventPayload.reason} 
+                  ${context.eventPayload.code} with isMonitoring set to ${this.keepAlive.isMonitoring} 
+                  and recommendReconnect set to ${this.keepAlive.recommendReconnect}`);
+                // this transition circumvents the 'disconnecting' state (since the websocket is already closed),
```
Comment (aoberoi) [inline src/RTMClient.ts]: ❤️  the attention to detail in the comments!

**Comment by aoberoi** [comment]
@shanedewael i know there's no testing rig for RTMClient at the moment, but it would be useful to know if you're done manual testing, and what that manual testing code might have looked like.

**Comment by shanedewael** [comment]
@aoberoi I manually hacked together a way to reproduce the bugs by emitting a handshake event in `RTMClient.ts`, listening to that event, and then forcibly closing the websocket in that state. This let me develop a solution to handle the error, and then by thinking through race conditions in `KeepAlive.ts`, prevent it from happening in the first place.

**Comment by vleipnik-consensys** [comment]
Guys, I am still getting an error with 4.2.0 
screen shot 2018-05-02 at 11 10 39 am

This seems related?

**Pull request merged**